plugin-server/src/main/ingestion-queues/session-recording/services/partition-locker.ts

```diff
@@ -0,0 +1,134 @@
+ */
+export class PartitionLocker {
+    consumerID = randomUUID()
+    delay = 1000
+    ttl = 30000
+
+    constructor(private redisPool: RedisPool, private keyPrefix = '@posthog/replay/locks') {}
+
+    private async run<T>(description: string, fn: (client: Redis) => Promise<T>): Promise<T> {
+        const client = await this.redisPool.acquire()
+        const timeout = timeoutGuard(`${description} delayed. Waiting over 30 seconds.`)
+        try {
+            return await fn(client)
+        } finally {
+            clearTimeout(timeout)
+            await this.redisPool.release(client)
+        }
+    }
+
+    private keys(tps: TopicPartition[]): string[] {
+        // Return a unique set of topicpartition keys
+        const keys = new Set<string>()
+        tps.forEach((tp) => keys.add(topicPartitionKey(this.keyPrefix, tp)))
+        return [...keys]
+    }
+    /* 
+        Claim the lock for partitions for this consumer
+        - If already locked, we extend the TTL
+        - If it is claimed, we wait and retry until it is cleared 
+        - If unclaimed, we claim it
+    */
+    public async claim(tps: TopicPartition[]) {
+        const keys = this.keys(tps)
+        const blockingConsumers = new Set(...[this.consumerID])
+
+        try {
+            while (blockingConsumers.size !== 0) {
+                blockingConsumers.clear()
+
+                await this.run(`claim keys that belong to this consumer`, async (client) => {
+                    await Promise.allSettled(
+                        keys.map(async (key) => {
+                            const existingClaim = await client.get(key)
+
+                            if (existingClaim && existingClaim !== this.consumerID) {
+                                // Still claimed by someone else!
+                                blockingConsumers.add(existingClaim)
+                                return
+                            }
+
+                            // Set the key so it is claimed by us
+                            const res = await client.set(key, this.consumerID, 'PX', this.ttl)
+                            if (!res) {
+                                blockingConsumers.add(this.consumerID)
+                            }
+                        })
+                    )
+                })
+
+                if (blockingConsumers.size > 0) {
+                    status.warn(
+                        '🔒',
+                        `PartitionLocker failed to claim keys. Waiting ${this.delay} before retrying...`,
+                        {
+                            id: this.consumerID,
+                            blockingConsumers,
+                        }
+                    )
+                    await new Promise((r) => setTimeout(r, this.delay))
+                }
+            }
+
+            status.info('🔒', 'PartitionLocker claimed all required keys')
+        } catch (error) {
+            status.error('🧨', 'PartitionLocker errored to claim keys', {
```

Comment:
Looking at the test failure... I guess we can't call `status` after the worker has exited but we're running async and might try to do things after exit...

Feels like a bug in status, if it isn't test setup/teardown